cpp/oneapi/dal/algo/knn/test/batch.cpp

```diff
@@ -0,0 +1,339 @@
+            std::iota(idcs_row, idcs_row + m, std::int32_t(0));
+            const auto compare = [&](std::int32_t x, std::int32_t y) -> bool {
+                return dist_row[x] < dist_row[y];
+            };
+            std::sort(idcs_row, idcs_row + m, compare);
+        }
+        return de::homogen_table_builder{}.reset(indices, n, m).build();
+    }
+
+    static auto arange(std::int64_t from, std::int64_t to) {
+        auto indices_arr = array<std::int32_t>::zeros(to - from);
+        auto* indices_ptr = indices_arr.get_mutable_data();
+        std::iota(indices_ptr, indices_ptr + to - from, std::int32_t(from));
+        return de::homogen_table_builder{}.reset(indices_arr, to - from, 1).build();
+    }
+
+    static auto arange(std::int64_t to) {
+        return arange(0, to);
+    }
+
+    void check_nans(const knn::infer_result<>& result) {
+        const auto [labels] = unpack_result(result);
+
+        SECTION("there is no NaN in labels") {
+            REQUIRE(te::has_no_nans(labels));
+        }
+    }
+
+    static auto unpack_result(const knn::infer_result<>& result) {
+        const auto labels = result.get_labels();
+        return std::make_tuple(labels);
+    }
+};
+
+using knn_types = COMBINE_TYPES((float, double), (knn::method::brute_force, knn::method::kd_tree));
+
+#define KNN_SMALL_TEST(name)                                               \
+    TEMPLATE_LIST_TEST_M(knn_batch_test,                                   \
+                         name,                                             \
+                         "[small-dataset][knn][integration][batch][test]", \
+                         knn_types)
+
+#define KNN_SYNTHETIC_TEST(name)                                               \
+    TEMPLATE_LIST_TEST_M(knn_batch_test,                                       \
+                         name,                                                 \
+                         "[synthetic-dataset][knn][integration][batch][test]", \
+                         knn_types)
+
+#define KNN_EXTERNAL_TEST(name)                                               \
+    TEMPLATE_LIST_TEST_M(knn_batch_test,                                      \
+                         name,                                                \
+                         "[external-dataset][knn][integration][batch][test]", \
+                         knn_types)
+
+KNN_SMALL_TEST("knn nearest points test predefined 7x5x2") {
+    SKIP_IF(this->not_available_on_device());
+
+    constexpr std::int64_t train_row_count = 7;
+    constexpr std::int64_t infer_row_count = 5;
+    constexpr std::int64_t column_count = 2;
+
+    CAPTURE(train_row_count, infer_row_count, column_count);
+
+    constexpr std::int64_t train_element_count = train_row_count * column_count;
+    constexpr std::int64_t infer_element_count = infer_row_count * column_count;
+
+    constexpr std::array<float, train_element_count> train = { -2.f, -1.f, -1.f,   -1.f,   -1.f,
```

Comment:
Use oneDAL array

Reply:
Such approach is used in `kmeans` tests i.e.